content/blog/sqlite_parser_pt_2.md

```diff
@@ -0,0 +1,529 @@
+will make a copy of the pages being changed and again put them in the journal file.
+The main difference is that it will not actually change the main database file but instead
+change the journal file. This means that any time someone wants to read information from
+the database, Sqlite first has to check if the data being requested is on a page that has changes
+in the journal before looking in the main database file. When all database connections
+are closed, or a configurable number of pages are modified, the changes are then
+applied to the main database file. This is scheme is sometimes referred to as 
+[append only](https://en.wikipedia.org/wiki/Append-only), by other databases.
+
+Now that we have covered what these next 2 bytes mean, let's setup our library to 
+parse them. Right now, Sqlite supports 2 values here, 1 for rollback journalling
+or "legacy mode" and 2 for write ahead log or WAL mode. The docs say that
+this could be extended to include additional values in the future. In our `header.rs`
+let's start with setting up an enum to represent this value.
+
+
+```rust
+// header.rs
+
+/// A value stored as a Write Format Version or
+/// Read Format Version
+#[derive(Debug, PartialEq, Eq, PartialOrd, Ord)]
+pub enum FormatVersion {
+    /// Represents the rollback journal mode
+    Legacy,
+    /// Represents the Write Ahead Log mode
+    WriteAheadLog,
+    /// Represents any mode not 1 or 2, the value
+    /// will be provided
+    Unknown(u8),
+}
+
+impl From<u8> for FormatVersion {
+    fn from(v: u8) -> Self {
+        match v {
+            1 => Self::Legacy,
+            2 => Self::WriteAheadLog,
+            _ => Self::Unknown(v),
+        }
+    }
+}
+```
+
+Let's just check in on a few things here. First we are `derive`ing a few more items,
+`PartialEq` and `Eq` allow for using the `==` operator with 2 `FormatVersion`s while
+the `PartialOrd` and `Ord` allow for the use of `>`, `<`, `<=`, or `>=`. We want to 
+go the extra mile here because the docs tell us that if the read format is 1 or 2
+but the write format is greater than 2 we have a read only database, when we get to 
+that validation it will be nice to use these operators.
+
+After we define our 3 possible values, we also can define `From<u8>` for our enum.
+If that value is 1 we return the `Legacy` mode, if 2 the `WriteAheadLog` otherwise
+we return the `Unknown` mode. This will allow for flexibility in parsing these values
+since the Sqlite authors may change the values available here moving forward. With
+our enum defined and that `From` implementation, it seems silly to add another
+parsing function since it will just wrap `FormatVersion::from`. Instead, let's
+add a main entry point for this module called `parse_header`.
+
+```rust
+// header.rs
+
+pub fn parse_header(bytes: &[u8]) -> Result<(PageSize, FormatVersion, FormatVersion), Error> {
+    // Check that the first 16 bytes match the header string
+    validate_header_string(&contents)?;
```

Comment:
this tripped me up at first but then I caught it, your references to `&contents` should actually be `&bytes` to match the input